### ISSUE TITLE
incorrect type checking in pattern matching

### ISSUE THREAD

**Issue by liufengyun** [lampepfl/dotty#1255]
Following code passes type checking in dotty:

``` Scala
object Test {
  def foo(x: Option[Int]) = x match {
    case Some(_: Double) => true
    case None => true
  }
}
```

In scalac it generates following error message:

```
error: scrutinee is incompatible with pattern type;
 found   : Double
 required: Int
    case Some(_: Double) => true
                 ^
one error found
```

**Comment by DarkDimius** [comment]
I propose to make it both more general and simpler by decoupling the transformation from typer and doing it after pattern matcher.

I propose to intercept `sc.isInstanceOf[Sel]`: depending on the erased type of `sc` and `Sel` either:
 - report the error\warning if test is always failing;
 - replace the call be constant if the result is statically known;
 - leave as is.

Here is the table for transformation:

| sel\Sc      | Trait                           | Class                           | final Class                |
|-------------|---------------------------------|---------------------------------|----------------------------|
| Trait       | happens                         | happens                         | known statically |
| Class       | happens                         | false if classes are unrelated | known statically |
| Final class | false if classes are unrelated |  false if classes are unrelated | known statically |

**Comment by DarkDimius** [comment]
I propose to make it both more general and simpler by decoupling the transformation from typer and doing it after pattern matcher.

I propose to intercept `sc.isInstanceOf[Sel]`: and, depending on the erased type of `sc` and `Sel`, either:
 - report the error\warning if test is always failing;
 - replace the call be constant if the result is statically known;
 - leave as is.
Here is the table for transformation:

| sel\Sc      | Trait                          | Class                          | final Class      |
|-------------|--------------------------------|--------------------------------|------------------|
| Trait       | happens                        | happens                | known statically |
| Class       | happens                        | always false if classes are unrelated | known statically |
| Final class | always false if classes are unrelated | always false if classes are unrelated | known statically |

**Comment by liufengyun** [comment]
@felixmulder I found an interesting bug with scalac related to this issue, which you may want to add to the test set:

```
case class A(v: Int)
class B(v: Int) extends A(v)

object Test {
  val a = new B(1)
  a match { case A(_) => 1 }
}
```

The code above looks perfectly fine, but scalac generates an error:

```
error: constructor cannot be instantiated to expected type;
 found   : A
 required: B
  a match { case A(_) => 1 }
                 ^
one error found
```

However, if we annote `a` with `val a:A = new B(1)`, everything is fine.

**Comment by felixmulder** [comment]
This issue was resolved in #1275
